src/web100srv.c

```diff
@@ -1951,375 +1837,335 @@ void NDT_server_main_loop(SSL_CTX *ssl_context, int listenfd) {
-
-        log_println(5, "Parent process spawned child = %d", chld_pid);
-        log_println(5, "Parent thinks pipe() returned fd0=%d, fd1=%d",
-                    chld_pipe[0], chld_pipe[1]);
-
-        close(chld_pipe[0]);
-
-        // Check to see if we have more than max_clients waiting in the
-        // queue
-        //  If so, tell them to go away.
-        //  changed for M-Lab deployment  1/28/09  RAC
-
-        if (((multiple == 0) && (waiting >= (max_clients - 1))) ||
-            ((multiple == 1) && (waiting >= ((4 * max_clients) - 1)))) {
-          log_println(
-              0,
-              "Too many clients/mclients (%d) waiting to be served, "
-              "Please try again later.", chld_pid);
-          send_json_message(ctlsockfd, SRV_QUEUE,
-            SRV_QUEUE_SERVER_BUSY_STR, 
-            strlen(SRV_QUEUE_SERVER_BUSY_STR),
-            testopt.connection_flags, JSON_SINGLE_VALUE);
-          close(chld_pipe[0]);
-          close(chld_pipe[1]);
-          shutdown(ctlsockfd, SHUT_WR);
-          close(ctlsockfd);
-          kill(chld_pid, SIGTERM);
-          if (new_child != NULL) {
-            log_println(6, "Too many clients freeing child=0x%x",
-                        new_child);
-            free(new_child);
-          }
-          continue;
+    chld_pid = fork();
+    if (chld_pid == -1) {
+      // an error occured, log it and keep waiting
+      log_println(0, "fork() failed, errno = %d (%s)", errno, strerror(errno));
+      close(child_pipe[0]);
+      close(child_pipe[1]);
+      close(ctlsockfd);
+      continue;
+    } else if (chld_pid == 0) {
+      // This is the child. Clean up the copies of the resources the child does
+      // not need, then call the child_process routine which will initialize the
+      // connection and then wait for the go/nogo signal from the parent
+      // process.
+      log_println(4, "Child thinks pipe() returned fd0=%d, fd1=%d for pid=%d",
+                  chld_pipe[0], chld_pipe[1], chld_pid);
+      close(listenfd);
+      close(chld_pipe[1]);
+      child_process(chld_pipe[0], ssl_context, ctlsockfd);
+      log_println(1, "The child returned! This should never happen.");
+      exit(1);
+    } else if (chld_pid == -1) {
+      // an error occured, log it and quit
+      log_println(0, "fork() failed, errno = %d", errno);
+      // todo: handle error and continue
+    } else {
+      // this is the parent process, handle scheduling and
+      // queuing of multiple incoming clients
+
+      log_println(5, "Parent process spawned child = %d", chld_pid);
+      log_println(5, "Parent thinks pipe() returned fd0=%d, fd1=%d",
+                  chld_pipe[0], chld_pipe[1]);
+
+      // Close the open resources that should only be used by the child.
+      close(chld_pipe[0]);
+      close(ctlsockfd);
+
+      // Check to see if we have more than max_clients waiting in the queue. If
+      // so, tell them to go away.
```

Comment:
This comment doesn't seem to be accurate. If `max_clients == 5` and `waiting == 4` we start killing children.